### ISSUE TITLE
RandomForests Performance Bug

### ISSUE THREAD

**Issue by AmeerHajAli** [dask/dask#5993]
Running randomforests with dask joblib backend does not scale well (4.8 minutes to run the below code). Multiprocessing performs 20X better. The machine used is m5.8xlarge instance. A code to reproduce:
```python
from joblib import parallel_backend
import numpy as np
from dask.distributed import Client
from sklearn.datasets import load_digits
from sklearn.ensemble import RandomForestClassifier
digits = load_digits()
from sklearn.model_selection import cross_val_score
client=Client(processes=None)
clf = RandomForestClassifier(n_estimators=45000,verbose=1)
X = np.concatenate((digits.data,digits.data),axis=0)
y = np.concatenate((digits.target,digits.target))
with parallel_backend('dask'):
    clf.fit(X,y)
```

**Comment by quasiben** [comment]
@AmeerHajAli  interesting. What is the performance if you pre-scatter the data:

```python
with parallel_backend('dask', scatter=[x, y]):
    search.fit(digits.data, digits.target)
```

**Comment by mrocklin** [comment]
I notice that you've chosen not to use processes:

```python
client=Client(processes=None)
```

Maybe this is the cause of the performance difference?  Maybe Scikit-Learn's RandomForest implementation doesn't release the GIL?  You might want to try with processes as a more comparable test.

If things are still slow then I recommend providing a performance report.  https://docs.dask.org/en/latest/diagnostics-distributed.html#capture-diagnostics

**Comment by AmeerHajAli** [comment]
Thanks for your responses!
@quasiben, I tried to use scatter, still the same issue.
@mrocklin, I tried `Client()` and `Client(processes=True)` and still does not work.
Do you suggest trying any other options?

Thanks!

**Comment by AmeerHajAli** [comment]
Just to make it clear, when I run it says:
"[Parallel(n_jobs=-1)]: Using backend DaskDistributedBackend with 32 concurrent workers."
But when I look at the utilization of the processor, it is less than 15%.

**Comment by TomAugspurger** [comment]
The Dask backend won't have any performance over scikit-learn on a single machine. If you aren't on a cluster you should just use the default for that estimator.

**Comment by AmeerHajAli** [comment]
@TomAugspurger, I also tried on a cluster with 5 and 10 nodes and it still does not perform well (the performance roughly stays the same). If instead of randomforests I do gridsearch on SVM hyperparameters it scales very well (on a single node and multiple nodes), but this is not the case for randomforests.

**Comment by TomAugspurger** [comment]
The predicate stage of RandomForestClassifier requires shared memory to mutate arrays inplace: https://github.com/scikit-learn/scikit-learn/blob/1c7449036515d4204f9467493010228588ed378d/sklearn/ensemble/_forest.py#L679-L683. So the fitting should be done in parallel on a cluster, but the CV evaluation parts might be done on your client machine.

https://github.com/joblib/joblib/issues/1020 might also be relevant, but given that you're seeing this just for random forest the shared memory requirement is the most likely culprit.

**Comment by AmeerHajAli** [comment]
Hi @TomAugspurger , I am not sure I understand, I am not using cross_validatiionn, it is just random forest with 45,000 trees. Does that mean that Dask performs significantly worse on random forests compared to multiprocessing? Can you suggest a parameter that I should pass to `parallel_backend` that will make it perform better?

**Comment by TomAugspurger** [comment]
Ah interesting, I misread that. I'd suggest looking at a performance report to see where time is spent.

> Can you suggest a parameter that I should pass to parallel_backend that will make it perform better?

Specifying `pre_dispatch="all"` may help alleviate https://github.com/joblib/joblib/issues/1020.

**Comment by mrocklin** [comment]
> @mrocklin, I tried Client() and Client(processes=True) and still does not work.

Yes.  This works fine for me.  Maybe if you're running this in a script you forgot to place process-creating code within an `if __name__ == "__main__":` block?

```python
...
if __name__ == "__main__":
    client = Client(...)
    with parallel_backend(...):
       ...
```

Looking at `top` while it's running I notice that my processes only use up to 100% CPU, despite each having a thread pool with a few cores each.  Probably the Scikit-Learn RandomForest code doesn't release the GIL, and so it is important to use processes rather than threads.  (You may want to read through https://docs.dask.org/en/latest/scheduling.html)

So probably you want something like the following ...

```python
client = Client(processes=True, threads_per_process=1)
```

**Comment by mrocklin** [comment]
Sorry, I meant `threads_per_worker`.  See the LocalCluster docstring for additional information.

**Comment by AmeerHajAli** [comment]
@mrocklin, did you try runnning the code with multiprocessing backend and dask and get similar performance? even with your fix, I still get dask being much slower than multiprocessing. This is my code:
```python
if __name__ == '__main__':
    from joblib import parallel_backend
    import numpy as np
    from dask.distributed import Client
    from sklearn.datasets import load_digits
    from sklearn.ensemble import RandomForestClassifier      
    digits = load_digits()
    from sklearn.model_selection import cross_val_score
    from time import time
    client=Client(processes=True,threads_per_worker=1)
    clf = RandomForestClassifier(n_estimators=45000,verbose=1)
    X = np.concatenate((digits.data,digits.data),axis=0)
    y = np.concatenate((digits.target,digits.target)) 
    with parallel_backend('dask'):
        start = time()
        clf.fit(X,y)
        end = time()
        print(end-start)
```

**Comment by mrocklin** [comment]
No, I didn't.

**Comment by AmeerHajAli** [comment]
Do you think it is a performance issue in Dask for RandomForests? Is it fine to assume Multiprocessing outperforms Dask in RandomForsts with Joblib backend?

**Comment by TomAugspurger** [comment]
When I run your example only two of the eight cores in my LocalCluster are being used. Do you see that on your task stream?

The dashboard or a performance report should help here.

**Comment by AmeerHajAli** [comment]
@TomAugspurger yes, that is the problem, no matter what configuration I try, it never utilizes 100% of the cores.

**Comment by TomAugspurger** [comment]
I'd recommend stepping through the code in
https://github.com/joblib/joblib/blob/master/joblib/_dask.py then to see
why. I don't recall if there are sufficient logging calls there to diagnose
things, so you may need to add print statements.

**Comment by AmeerHajAli** [comment]
But just to make sure, this issue is not only on my machine. You (@TomAugspurger , @mrocklin) also cannot get it to work on your machines, right?

**Comment by TomAugspurger** [comment]
I saw only 2/8 cores being used.

**Comment by mrocklin** [comment]
Tom, I think that random forest code holds on to the GIL.  My guess is that
your local cluster is using two processes with four threads each.

When I run with N processes I see all processes active.

**Comment by AmeerHajAli** [comment]
@mrocklin, do you mean to set `n_workers=N`? Because even that does not work.
Can you please try running with multiprocessing and dask and report the performance?

Thanks!

**Comment by mrocklin** [comment]
I meant that RandomForest holds onto the GIL, and so one probably wants to
have many processes with one thread each.

**Comment by mrocklin** [comment]
> Can you please try running with multiprocessing and dask and report the performance?

I don't know how to try running your code with multiprocessing unfortunately.  If you have something that I can copy-paste then sure, I'd be happy to run it on my machine.

Also, if you're looking for help with understanding why something is slow, then I recommend providing a performance_report .  See https://docs.dask.org/en/latest/understanding-performance.html

**Comment by AmeerHajAli** [comment]
Just replace dask with multiprocessing, here:
```python
if __name__ == '__main__':
    from joblib import parallel_backend
    import numpy as np
    from dask.distributed import Client
    from sklearn.datasets import load_digits
    from sklearn.ensemble import RandomForestClassifier      
    digits = load_digits()
    from sklearn.model_selection import cross_val_score
    from time import time
    //client=Client(processes=True,threads_per_worker=1)
    clf = RandomForestClassifier(n_estimators=45000,verbose=1)
    X = np.concatenate((digits.data,digits.data),axis=0)
    y = np.concatenate((digits.target,digits.target)) 
    with parallel_backend('multiprocessing'):
        start = time()
        clf.fit(X,y)
        end = time()
        print(end-start)
```

**Comment by mrocklin** [comment]
I went ahead and generated the performance report

https://gist.githack.com/mrocklin/de1efe90a59716323d32dd4b492c50a4/raw/f8e6e24f0e61a1763ef454155d4010979d2e2ed6/random-forest.html

We're spending all of our time in some low-level `funlockfile` function.  Perhaps there is some contention with some lower level resource on the Scikit-Learn side?  If you have time to investigate this @AmeerHajAli that would be very much welcomed.

**Comment by mrocklin** [comment]
cc @ogrisel and @lesteve , who might be interested in this

**Comment by mrocklin** [comment]
Looking at the dashboard when it's running it appears that there are not enough tasks in flight and the workers aren't sufficiently occupied.  Does anyone know where the code within joblib is that determines how many futures we have active at once?

**Comment by lesteve** [comment]
@TomAugspurger is right about:

> The predicate stage of RandomForestClassifier requires shared memory to mutate arrays inplace: https://github.com/scikit-learn/scikit-learn/blob/1c7449036515d4204f9467493010228588ed378d/sklearn/ensemble/_forest.py#L679-L683

In other words, multi-threading is used when doing RandomForest without Dask so I don't expect the scikit-learn parallelized function to hold to the GIL.

I don't really know what happens with `require="sharedmem"` when passed to the Dask backend. Maybe that gets turned i_nto maybe it runs the code only one worker. Unless I am mistaken 8 cores turns into 4 process and 2 cores:
```
In [1]: from distributed.deploy.local import nprocesses_nthreads                                                                     

In [2]: nprocesses_nthreads(8)                                                                                                       
Out[2]: (4, 2)
```

So that would make sense for the 2 cores used reported above

Also this ay be related to a similar performance problem in https://github.com/joblib/joblib/issues/1021.

**Comment by mrocklin** [comment]
So I'm mostly concerned about the sparsity of tasks during computation

**Comment by mrocklin** [comment]
Whoops, I meant to include some more information there.  This is a screenshot from the [performance report](https://gist.githack.com/mrocklin/de1efe90a59716323d32dd4b492c50a4/raw/f8e6e24f0e61a1763ef454155d4010979d2e2ed6/random-forest.html)

![task-stream](https://user-images.githubusercontent.com/306380/77236632-a31ea500-6b7d-11ea-918d-c20904ea0a55.png)

Joblib isn't feeding Dask with enough concurrent work.  When I watch the dashboard during execution I see that there are only about 20 tasks total, 16 of which are done most of the time, so there is only a backlog of something like 4 tasks at once, which doesn't saturate the number of cores I have (I would want something like twice the number of total cores for saturation).  

This this specific to RandomForest, or is this a general Joblib+Dask problem?  Where does the amount of concurrency get set?

**Comment by TomAugspurger** [comment]
https://github.com/joblib/joblib/issues/1020 for a bit about joblib not submitting enough tasks. No concrete conclusions yet.

**Comment by lesteve** [comment]
For a general view about joblib from one of Gäel's [talk](https://www.slideshare.net/GaelVaroquaux/processing-biggish-data-on-commodity-hardware-simple-python-patterns):
![image](https://user-images.githubusercontent.com/1680079/77236935-77261280-6bc3-11ea-9ad9-496218613d1c.png)

My general understanding (it has been a while since I haven't looked at the code in detail):
- the `pre_dispatch` puts some tasks in the task queue before any computation starts. The default `2 * n_jobs` is chosen so that the workers always have something to do
- the size of the (process or thread) pool is `n_jobs` (argument passed to `Parallel`)
- when a task finishes, there is a callback that pushes the next task to the task queue

On top of that there is some auto-batching feature, that is supposed to batch tasks automatically to reach a target time (especially useful when the task is too short because of serialisation and communication overhead). That does not seem to kick in the Dask case for some reason if I read the dashboard correctly (or the Joblib task time does not represent the actual computation time accurately reported by Dask).

**Comment by mrocklin** [comment]
I pushed up https://github.com/joblib/joblib/pull/1025 , which helps a bit.  There is still plenty of dead time though.  I'm not entirely sure where.

**Comment by Chroxvi** [comment]
As the submitter of https://github.com/joblib/joblib/issues/1021, I'll chime in with my experiences on issues with the combination of `RandomForestClassifier` and the dask backend for joblib. I think we are dealing with at least 3 separate issues here:

1. When fitting (and possibly also when predicting) a `RandomForestClassifier` using the dask backend for joblib, in my experience, you need to use processes in order to have everything parallelize nicely. I haven't done any digging to confirm if this it due to the GIL not being released as suggested by @mrocklin.
2. When doing predictions for a `RandomForestClassifier` using the dask backend for joblib, the `require="sharedmem"` seems to be ignored. As a result all the prediction tasks for the individual estimators are submitted to the dask scheduler at once generally resulting in a *HUGE* memory usage as the entire array of probabilities of size (n_samples, n_classes, n_estimatores) from `predict_proba` is kept in memory.
3. When fitting a `RandomForestClassifier` using the dask backend for joblib, it seems to parallelize nicely for small jobs, i.e. few features, few samples, few classes, few estimators. However, for larger problems, all tasks end up being scheduled on a single worker as described in https://github.com/joblib/joblib/issues/1021. I strongly suspect that this has to do with the way joblib submits tasks to dask. If I bypass joblib and directly submit tasks to dask (see ugly monkey patch in https://github.com/joblib/joblib/issues/1021#issuecomment-602586791), I am able to keep the dask cluster saturated. Maybe this is the same problem @AmeerHajAli is experiencing?

Maybe the last problem is related to https://github.com/joblib/joblib/issues/1020. I am having a hard time finding my way around the interplay of dask, joblib, and `RandomForestClassifier`. I'd be interested in doing some more digging if anyone can point me to the possible cause of the last problem.

**Comment by mrocklin** [comment]
> When doing predictions for a RandomForestClassifier using the dask backend for joblib, the require="sharedmem" seems to be ignored. As a result all the prediction tasks for the individual estimators are submitted to the dask scheduler at once generally resulting in a HUGE memory usage as the entire array of probabilities of size (n_samples, n_classes, n_estimatores) from predict_proba is kept in memory

This would explain what I'm seeing.  It's taking a while for joblib to collect results from Dask.  If those results were very large that would explain things.  

I don't know ML algorithms well, but I assumed that RandomForest was relatively embarrassingly parallel.  I take it that there is some rapidly changing state that is important to share?

**Comment by Chroxvi** [comment]
> I don't know ML algorithms well, but I assumed that RandomForest was relatively embarrassingly parallel. I take it that there is some rapidly changing state that is important to share?

Training and predicting using a RandomForest is a very embarrasingly parallel problem. Basically you train a number (n_estimators) of individual decision trees and then you take the majority vote (for classification) or average (for regression) of those individual trees when doing prediction. There is a little more to it than that, but from a computational perspective, you can treat the training of the individual trees as independent of each other and the prediction using the individual trees as independent of each other.

The prediction (`predict_proba` which is used in `predict`) using joblib for the scikit-learn RandomForestClassifier happens here: https://github.com/scikit-learn/scikit-learn/blob/b194674c4/sklearn/ensemble/_forest.py#L665 and uses this `_accumulate_prediction` function https://github.com/scikit-learn/scikit-learn/blob/b194674c4/sklearn/ensemble/_forest.py#L442.

As far as I understand, when you pass the `require="sharedmem"` to joblib, it forces the use of a thread based backend that computes everything sequentially and allows for shared memory.  All the individual trees in the RandomForestClassifier produce an output array of sizes (n_samples, n_classes) which must be summed and normalized. The summation happens inplace in a preallocated array in `_accumulate_prediction`. Thus, all of this seems to happen to avoid having all n_estimators arrays of size (n_samples, n_classes) in memory at the same time.

It seems to me, that this "sharedmem" constraint is ignore when using the dask backend for joblib. I haven't looked into the details - only experiences some *HUGE* memory usage during prediction that I can't otherwise explain.

**Comment by mrocklin** [comment]
Yeah, so I guess my question becomes, "If RandomForest is embarrassingly parallel then why are we sending around a lot of data?"

**Comment by ogrisel** [comment]
> It seems to me, that this "sharedmem" constraint is ignore when using the dask backend for joblib. I haven't looked into the details - only experiences some HUGE memory usage during prediction that I can't otherwise explain.

That would be bug of its own.

The `fit` call should be both embarrassingly parallel and not require sharedmem, hence should be able to benefit from dask work level parallelism. Maybe the dask workers are starving because the pre-dispatch logic is somehow broken with the dask backend.

I know @pierreglaser started to run various joblib/dask benchmarks and also observed this in some case.

**Comment by Chroxvi** [comment]
> The `fit` call should be both embarrassingly parallel and not require sharedmem, hence should be able to benefit from dask-worker level parallelism. Maybe the dask workers are starving because the pre-dispatch logic is somehow broken with the dask backend.

Just to be clear. My comments about sharedmem relate to prediction (the `RandomForestClassifier.predict()` and `RandomForestClassifier.predict_proba()` methods) not fitting (the `RandomForestClassifier.fit()` method) using a `RandomForestClassifier`.

**Comment by martindurant** [comment]
Can we mark this as at least partially resolved?

For a discussion on the status of shared memory in one-node distributed, see https://github.com/dask/distributed/issues/4497
